File: common/sw/architecture/storage_hierarchy.py
from abc import ABC

class StorageArchitecture(ABC):
    """
    Base class for all computer architectures in the zoo.
    Defines the fundamental parameters of the design space.
    """
    word_width: int
    address_width: int

class StackArchitecture(StorageArchitecture):
    """
    Stack Architecture (zero-address storage hierarchy).
    Operands are implicitly on top of a stack.
    """
    stack_depth: int

class AccumulatorArchitecture(StorageArchitecture):
    """
    Accumulator Architecture (single-address storage hierarchy).
    Operands are implicitly in a single special register (Accumulator).
    """
    has_link_bit: bool

class GeneralRegisterArchitecture(StorageArchitecture):
    """
    General-Purpose Register (GPR) Architecture (double/multi-address storage).
    Operands are stored in a file of addressable registers.
    """
    num_gprs: int
    num_fprs: int

class VectorArchitecture(GeneralRegisterArchitecture):
    """
    Vector Register Architecture.
    Support for vector execution on arrays of data.
    """
    vector_length: int
    num_vec_regs: int

File: common/rtl/src/main/scala/zoo/common/architecture/StorageHierarchy.scala
package zoo.common.architecture

import chisel3._

/**
 * Base trait for all computer architectures in the zoo.
 * Defines the fundamental parameters of the design space.
 */
trait StorageArchitecture {
  val wordWidth: Int
  val addressWidth: Int
}

/**
 * Stack Architecture (zero-address storage hierarchy).
 * Operands are implicitly on top of a stack.
 * Example: Burroughs B5500.
 */
trait StackArchitecture extends StorageArchitecture {
  val stackDepth: Int
}

/**
 * Accumulator Architecture (single-address storage hierarchy).
 * Operands are implicitly in a single special register (Accumulator).
 * Example: DEC PDP-8.
 */
trait AccumulatorArchitecture extends StorageArchitecture {
  val hasLinkBit: Boolean // E.g., PDP-8 has a 1-bit Link register for carry
}

/**
 * General-Purpose Register (GPR) Architecture (double/multi-address storage).
 * Operands are stored in a file of addressable registers.
 * Example: IBM System/360, Cray-1.
 */
trait GeneralRegisterArchitecture extends StorageArchitecture {
  val numGPRs: Int
  val numFPRs: Int // Floating point registers
}

/**
 * Vector Register Architecture.
 * Support for vector execution on arrays of data.
 * Example: Cray-1.
 */
trait VectorArchitecture extends GeneralRegisterArchitecture {
  val vectorLength: Int // Number of elements per vector register
  val numVecRegs: Int   // Number of vector registers
}
